packages/react-components/src/components/MessageThread.tsx

```diff
@@ -572,6 +574,20 @@ export type MessageThreadProps = {
    * Optional strings to override in component
    */
   strings?: Partial<MessageThreadStrings>;
+
+  /* @conditional-compile-remove(file-sharing) */
+  /**
+   * @beta
+   * Optional function called when someone clicks on the file download icon.
+   */
+  fileDownloadHandler?: FileDownloadHandler;
+  /* @conditional-compile-remove(file-sharing) */
+  /**
+   * @beta
+   * Property name that contains information about file downloads in `message.metadata` object.
+   * @defaultValue fileSharingMetadata
+   */
+  fileDownloadMetadataKey?: string;
```

Comment:
Could this be different per message? Or guaranteed to be consistent across the whole message thread?

Reply:
Consistent for the whole message thread since this prop is set at the message thread level

Reply:
Might Contoso want a different key for different messages? Struggling to think of a good reason why they would do this.. perhaps they updated to a new key as the json structure changed but want to support old key for backwards compat reasons? Just don't want to limit the API today to one `fileDownloadMetadataKey` per message thread and find out we need one per message in the future

Reply:
Fair point. One way to do it would be to embed the key information in `message.metadata` itself.
```tsx
metadata = {
  fileDownloadMetadataKey: "fileDownloads",
  fileDownloads: []
}
```

Reply:
But this makes the concept much more harder for Contoso to grasp.

Reply:
@PorterNan @prprabhu-ms thoughts?

Reply:
Would also love to get @PorterNan or @prprabhu-ms input here, my feeling is a download is contained inside a message so I feel it make more sense for the message props to encapsulate the key to retrieve the download from the message data.
I think I would have it at the `ChatMessage` interface level:
```tsx
interface ChatMessage extends MessageCommon {
  messageType: 'chat';
  content?: string;
  contentType: MessageContentType;
  ...
  metadata?: Record<string, string>;
  // key inside the metadata field that houses file attachment information
  fileMetadataKey: string;

  // in theory other attachments could then be added:
  powerpointMetadatKey; // specifically a onedrive hosted powerpoint link to embed a live preview in message
  arcadeMetadata; // information on a arcade component that is to be live embedded in a chat message
  // ^ though perhaps at this point its not a ChatMessage and its something larger 
}
```

Reply:
In contrast to changing the interface.
I would recommend storing the `fileMetadataKey` in the `message.metadata` itself.
@JamesBurnside keys other than `metadata` won't be persisted with the message. The info will be lost when the message is sent...

Reply:
There's going to have to be a contract in place in Contoso's application that defines what the metadata key is regardless, so I think that information is lose-able when messages are sent.

If the fileMetadataKey is stored in message.metadata, the receiver will need to know to lookup `fileMetadataKey` as a fixed prop anyway (so shifts the contract to being defined by us instead of consoto, at which point we may as well have a fixed fileMetadataKey that is always called fileMetadataKey and not let contoso use a custom key)

Reply:
Is this file download metadata syntax defined by us or our chat sdk?

If that is defined by us, it would be great to add a "type" property into the metadata type, and do some typescript magic
If type="fileDownload" then we limit the metadata to be file metadata type, so we are good to put the key into metadata

Reply:
@PorterNan I love that idea. It can be done.

Reply:
> How does that handle the scenario where each message has different key for attached files if the contract is statically defined in Contoso's application?

Not a hardcoded constant, just a contract between contoso applications. There could be if-else's in the code to choose different keys, or perhaps a service updates that key, but agree that is a contrived example.

I think I see message thread a our container for chat messages, it arranges and styles messages, but chat messages are isolated components that should be able to exist outside of a message thread. We don't export a ChatMessage component so we don't have to support that, but in terms of component encapsulation it should be possible to have a chat message with a file download elsewhere. Imagine if we implemented the full chat notifications and allowed a download button inside one of these:
![image](https://user-images.githubusercontent.com/2684369/163262363-5f562b84-e2f9-4f9f-966b-1625453e0137.png)